cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java

```diff
@@ -0,0 +1,145 @@
+  private static final Gson GSON = new Gson();
+  private static final Type TOKEN_TYPE_SET_STRING = new TypeToken<Set<String>>() {
+  }.getType();
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new com.google.common.reflect.TypeToken<Map<String,
+    MonitorConsumeRequest>>() {
+  }.getType();
+  private static final JsonParser JSON_PARSER = new JsonParser();
+
+  private final CConfiguration cConf;
+  private final MessageFetcher messageFetcher;
+
+  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher) {
+    this.cConf = cConf;
+    this.messageFetcher = messageFetcher;
+  }
+
+  /**
+   * Returns map of topic config -> topic name to be monitored
+   */
+  @POST
+  @Path("/monitor/topics")
+  public void topics(FullHttpRequest request, HttpResponder responder) throws Exception {
+    try {
+      String data = request.content().toString(StandardCharsets.UTF_8);
+
+      if (!isValidJSON(data)) {
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Invalid JSON in body");
+        return;
+      }
+
+      Set<String> topics = GSON.fromJson(data, TOKEN_TYPE_SET_STRING);
+      Map<String, String> returnMap = new HashMap<>();
+
+      for (String topic : topics) {
+        returnMap.put(topic, cConf.get(topic));
+      }
+
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(returnMap));
+    } catch (Exception ex) {
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error: " + ex.getMessage());
+    }
+  }
+
+  /**
+   * Returns runtime related metadata.
+   */
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    try {
+      String data = request.content().toString(StandardCharsets.UTF_8);
+
+      if (!isValidJSON(data)) {
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Invalid JSON in body");
+        return;
+      }
+
+      Map<String, MonitorConsumeRequest> topicsToFetch = GSON.fromJson(data, MAP_STRING_CONSUME_REQUEST_TYPE);
+      Map<String, List<MonitorMessage>> returnMap = new HashMap<>();
+
+      for (Map.Entry<String, MonitorConsumeRequest> topicToFetch : topicsToFetch.entrySet()) {
+        List<MonitorMessage> list = new LinkedList<>();
+        try (CloseableIterator<Message> iter = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    cConf.get(topicToFetch.getKey()),
+                                                                    topicToFetch.getValue().getLimit(),
+                                                                    topicToFetch.getValue().getMessageId())) {
+          while (iter.hasNext()) {
+            Message message = iter.next();
+            list.add(new MonitorMessage(message.getId(), message.getPayloadAsString(StandardCharsets.UTF_8)));
+          }
+        }
+
+        returnMap.put(topicToFetch.getKey(), list);
```

Comment:
To have better memory efficiency, better use `JsonWriter` to stream the result back. You can use the `responder.sendChunkStart` to help